sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/V2Aggregator.scala

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions.aggregate
+
+import java.io.{ByteArrayInputStream, ByteArrayOutputStream, ObjectInputStream, ObjectOutputStream}
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.Expression
+import org.apache.spark.sql.connector.catalog.functions.{AggregateFunction => V2AggregateFunction}
+import org.apache.spark.sql.types.DataType
+
+case class V2Aggregator[BUF <: java.io.Serializable, OUT](
+    aggrFunc: V2AggregateFunction[BUF, OUT],
+    children: Seq[Expression],
+    mutableAggBufferOffset: Int = 0,
+    inputAggBufferOffset: Int = 0) extends TypedImperativeAggregate[BUF] {
+  override def createAggregationBuffer(): BUF = aggrFunc.newAggregationState()
+
+  override def update(buffer: BUF, input: InternalRow): BUF = aggrFunc.update(buffer, input)
```

Comment:
This seems incorrect. The actual input is `children.map(_.eval(input))`, which is not an `InternalRow` and the "magic method" is useful in `V2AggregateFunction` as well.

Reply:
Similar to `ScalarFunction`, we can document that `invokeUpdate` is the magic method name.

Reply:
Oops you are right. We should evaluate the `input` first. I'll also add a test case for this.

Let me also handle the magic method case for aggregate functions (was planning to do that separately).

Reply:
@cloud-fan I checked the magic method support for aggregate functions and there are a few issues:
1. seems we can't reuse the existing `Invoke` as the `invokeUpdate` method will also take aggregate state `BUF` as input, while `Invoke` evaluates on `InternalRow` only.
2. only hash aggregation support codegen at the moment and also it doesn't support `ImperativeAggregate`.

Do you mind we leave this as a TODO for now and handle it separately? I think it requires some work to implement this.

Reply:
SGTM, let's wrap `children.map(_.eval(input))` into an `InternalRow` and more forward first.